README.txt
= libmagic

* git://git.collaborativedrug.com/libmagic.git

== DESCRIPTION:

Ruby wrapper for the Unix file/libmagic utility, which can guess mime types and character sets.

== FEATURES/PROBLEMS:

* Determines mime type of a file, IO or string
* Can also just return the charset of a text file
* Cannot detect various flavors of extended ASCII that are not ISO-8859-1, returns "unknown" as the charset in all cases
* For many uses this is good enough, because by far the most common "unknown" will be windows-1252

== SYNOPSIS:

  Magic.file_mime_type("path_to_file.pdf")
  Magic.file_charset("path_to_file.txt")

== REQUIREMENTS:

* the Unix program file (aka libmagic)'s libraries and headers (see below for installation)

== INSTALL:

* install file/libmagic's libraries and headers
  * on Ubuntu: sudo aptitude install libmagic-dev
  * on OS X with MacPorts: sudo port install file
    Unfortunately, the version of file that comes with OS X appears to be statically linked

* sudo gem install libmagic

== LICENSE:

(The MIT License)

Copyright (c) 2008 Collaborative Drug Discovery, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG
# Version 0.5.11
* Address FFI deprecation errors with ruby 2.7
# Version 0.5.6
* Fix regression in Magic#string_charset for an ASCII string
# Version 0.5.5
* More robust detection of problematic ellipses in windows-1252 files, such that it doesn't get confused by utf-8 files with angstroms in them
# Version 0.5.4
* Now supports Magic.io_charset(io), which will determine the charset of an IO by reading the whole thing
# Missing some stuff in here...
# Version 0.2
* now working on Linux! (libmagic returns different strings for magic_file with the MAGIC_MIME option on the two platforms)
# Version 0.1
* initial version, works on OS X


test/test_magic.rb
# coding: utf-8
require "rubygems"
require "lib/libmagic"
require "test/unit"
require "stringio"

class MagicTest < Test::Unit::TestCase
  def test_public_interface_is_limited
    assert_equal(%w(file_charset file_charset! file_mime_type io_charset string_charset string_mime_type).map { |m| m.to_sym },
                 (Magic.public_methods - Magic.instance_methods - FFI::Library.methods - FFI::Library.instance_methods).sort.map { |m| m.to_sym })
  end
  
  def test_file_mime_type_for_utf8_file
    # regex necessary because some versions of file return the semicolon and some don't
    assert(Magic.file_mime_type(absolute_path("utf-8.txt")) =~ /text\/plain;? charset=utf-8/)
  end
  
  def test_string_mime_type_for_utf8_text
    # regex necessary because some versions of file return the semicolon and some don't
    assert(Magic.string_mime_type("Some truly Unicode characters like: 불거기") =~ /text\/plain;? charset=utf-8/)
  end
  
  def test_string_charset_for_ascii_text
    assert_equal("us-ascii", Magic.string_charset("Just ASCII"))
  end
  
  def test_string_charset_for_utf8_text
    assert_equal("utf-8", Magic.string_charset("Some truly Unicode characters like: 불거기"))
  end
  
  def test_string_charset_for_iso_8859_1_text
    assert_equal("iso-8859-1", Magic.string_charset("A\240B\240C"))
  end
  
  (128..159).each do |windows_char|
    eval <<-EOMETHOD
      def test_string_charset_for_string_with_windows_char_#{windows_char}_returns_unknown
        assert_equal("unknown", Magic.string_charset("over, and over, and over\\#{windows_char.to_s(8)}"))
      end
    EOMETHOD
  end
  
  # although this is redundant, it's nice to document the one weird case we've found explicitly
  def test_string_charset_for_string_with_windows_ellipsis_returns_unknown
    # windows 1252 ellipsis is 133 = 0205
    # this is a weird case found from a file Sylvia had problem slurping
    assert_equal("unknown", Magic.string_charset("over, and over, and over\205"))
  end
  
  def test_string_charset_for_string_with_utf8_Angstrom_returns_utf8_not_unknown
    string = File.open(absolute_path("utf-8.csv.gz")) do |io|
      uncompressed_io = Zlib::GzipReader.new(io)
      uncompressed_io.read
    end
    assert_equal("utf-8", Magic.string_charset(string))
  end
  
  def test_io_charset_for_ascii_file
    assert_equal("us-ascii", Magic.io_charset(File.open((absolute_path("us-ascii.txt")))))
  end
  
  def test_io_charset_for_utf8_file
    assert_equal("utf-8", Magic.io_charset(File.open(absolute_path("utf-8.txt"))))
  end
  
  require "zlib"
  def test_io_charset_for_gzipped_utf8_file
    File.open(absolute_path("utf-8.csv.gz")) do |io|
      uncompressed_io = Zlib::GzipReader.new(io)
      assert_equal("utf-8", Magic.io_charset(uncompressed_io))
    end
  end
  
  def test_io_charset_for_iso_8859_1_file
    assert_equal("iso-8859-1", Magic.io_charset(File.open(absolute_path("iso-8859-1.txt"))))
  end
  
  def test_io_charset_for_windows_1252_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.io_charset(File.open(absolute_path("windows-1252.txt"))))
  end
  
  def test_io_charset_for_macintosh_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.io_charset(File.open(absolute_path("macintosh.txt"))))
  end
  
  def test_io_charset_for_csv_file_that_looked_like_ppm_image
    assert_equal("us-ascii", Magic.io_charset(File.open(absolute_path("file_with_text_that_looked_like_ppm_image.csv"))))
  end
  
  def test_file_charset_for_ascii_file
    assert_equal("us-ascii", Magic.file_charset(absolute_path("us-ascii.txt")))
  end
  
  def test_file_charset_for_large_CSV_file_that_libmagic_thinks_is_pascal_sourcecode
    assert_equal("us-ascii", Magic.file_charset(absolute_path("part_of_ki_file.csv")))
  end
  
  def test_file_charset_for_utf8_file
    assert_equal("utf-8", Magic.file_charset(absolute_path("utf-8.txt")))
  end
  
  def test_file_charset_for_iso_8859_1_file
    assert_equal("iso-8859-1", Magic.file_charset(absolute_path("iso-8859-1.txt")))
  end
  
  def test_file_charset_for_windows_1252_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.file_charset(absolute_path("windows-1252.txt")))
  end
  
  def test_file_charset_for_macintosh_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.file_charset(absolute_path("macintosh.txt")))
  end
  
  def test_file_charset_for_csv_file_that_looked_like_ppm_image
    assert_equal("us-ascii", Magic.file_charset(absolute_path("file_with_text_that_looked_like_ppm_image.csv")))
  end
  
  def test_file_charset_raises_if_file_does_not_exist
    # for this, we don't use assert_raise
    begin
      Magic.file_charset("some file that does not exist.txt")
      fail "Did not raise"
    rescue Exception => expected
      # ruby 1.9 and 1.8 return different exceptions
      assert(expected.message =~ /(some file that does not exist.txt|NULL pointer)/i)
    end
  end
  
  def test_file_charset_bang_exhaustively_checks_file_contents
    # huge_file_with_one_special_character.csv
    t1 = Time.now
    assert_equal("us-ascii", Magic.file_charset(absolute_path("huge_file_with_one_special_character.csv")))
    assert_equal("iso-8859-1", Magic.file_charset!(absolute_path("huge_file_with_one_special_character.csv")))
    puts "took #{Time.now - t1} seconds"
  end
  
  def test_file_charset_bang_returns_correct_value_for_us_ascii_file
    assert_equal("us-ascii", Magic.file_charset!(absolute_path("us-ascii.txt")))
  end
  
  def test_file_charset_bang_returns_correct_value_for_windows_1252_file
    assert_equal("unknown", Magic.file_charset!(absolute_path("windows-1252.txt")))
  end
  
  def test_file_charset_bang_returns_correct_value_for_UTF8_file
    assert_equal("utf-8", Magic.file_charset!(absolute_path("utf-8.txt")))
  end
  
  def test_file_charset_bang_handles_special_character_at_the_end_of_the_file
    assert_equal("us-ascii", Magic.file_charset(absolute_path("huge_file_with_one_special_character_at_the_end.csv")))
    assert_equal("iso-8859-1", Magic.file_charset!(absolute_path("huge_file_with_one_special_character_at_the_end.csv")))
  end
  
  def test_collect_special_characters_is_empty_when_there_are_no_special_characters
    assert_special_chars_equal("", "")
    assert_special_chars_equal("", "hello")
    assert_special_chars_equal("", "12345678901234567890")
  end
  
  def test_collect_special_characters_returns_characters_with_context
    assert_special_chars_equal("µ", "µ")
    assert_special_chars_equal("321µ123", "321µ123")
    assert_special_chars_equal("µ123", "µ123")
    assert_special_chars_equal("321µ", "321µ")
    assert_special_chars_equal("0987654321\xC21234567890", "0987654321\xC21234567890")
    assert_special_chars_equal("0987654321µ1234567890", "XXX0987654321µ1234567890XXX")
    assert_special_chars_equal("µ1234567890", "µ1234567890XXX")
    assert_special_chars_equal("0987654321µ", "XXX0987654321µ")
  end
    
  def test_collect_special_characters_does_not_duplicate_context
    assert_special_chars_equal("0987654321µaaaaaµ1234567890", "XXX0987654321µaaaaaµ1234567890XXX")
  end
  
  def test_collect_special_characters_works_with_multiple_characters
    assert_special_chars_equal(
      "0987654321µaaaaaµ12345678900987654321µ1234567890", 
      "XXX0987654321µaaaaaµ1234567890XXXXXX0987654321µ1234567890XXX"
    )
  end
  
  def test_collect_special_characters_works_when_reading_multiple_chunks_to_the_buffer
    default_chunk_size = 2 ** 15
    assert_equal(default_chunk_size, Magic::CHUNK_SIZE)
    begin
      Magic.send(:remove_const, "CHUNK_SIZE")
      Magic.send(:const_set, "CHUNK_SIZE", 2)
      assert_special_chars_equal("µ", "µ")
      assert_special_chars_equal("µ123", "µ123")
      assert_special_chars_equal("321µ", "321µ")
      assert_special_chars_equal("321µ123", "321µ123")
      assert_special_chars_equal("0987654321µ", "XXX0987654321µ")
      assert_special_chars_equal(
        "0987654321µaaaaaµ12345678900987654321µ1234567890",
        "XXX0987654321µaaaaaµ1234567890XXXXXX0987654321µ1234567890XXX"
      )
    ensure
      Magic.send(:remove_const, "CHUNK_SIZE") if Magic.const_defined?("CHUNK_SIZE")
      Magic.const_set("CHUNK_SIZE", default_chunk_size)
    end
    assert_equal(default_chunk_size, Magic::CHUNK_SIZE)
  end
  
  def assert_special_chars_equal(expected_output, input)
    assert_equal(expected_output.force_encoding(Encoding::BINARY), Magic.send(:collect_special_characters, StringIO.new(input)))
  end
  
  def absolute_path(test_file_name)
    "#{ENV["PWD"]}/test/files/#{test_file_name}"
  end
end


pkg/libmagic-0.5.11/test/test_magic.rb
# coding: utf-8
require "rubygems"
require "lib/libmagic"
require "test/unit"
require "stringio"

class MagicTest < Test::Unit::TestCase
  def test_public_interface_is_limited
    assert_equal(%w(file_charset file_charset! file_mime_type io_charset string_charset string_mime_type).map { |m| m.to_sym },
                 (Magic.public_methods - Magic.instance_methods - FFI::Library.methods - FFI::Library.instance_methods).sort.map { |m| m.to_sym })
  end

  def test_file_mime_type_for_utf8_file
    # regex necessary because some versions of file return the semicolon and some don't
    assert(Magic.file_mime_type(absolute_path("utf-8.txt")) =~ /text\/plain;? charset=utf-8/)
  end

  def test_string_mime_type_for_utf8_text
    # regex necessary because some versions of file return the semicolon and some don't
    assert(Magic.string_mime_type("Some truly Unicode characters like: 불거기") =~ /text\/plain;? charset=utf-8/)
  end

  def test_string_charset_for_ascii_text
    assert_equal("us-ascii", Magic.string_charset("Just ASCII"))
  end

  def test_string_charset_for_utf8_text
    assert_equal("utf-8", Magic.string_charset("Some truly Unicode characters like: 불거기"))
  end

  def test_string_charset_for_iso_8859_1_text
    assert_equal("iso-8859-1", Magic.string_charset("A\240B\240C"))
  end

  (128..159).each do |windows_char|
    eval <<-EOMETHOD
      def test_string_charset_for_string_with_windows_char_#{windows_char}_returns_unknown
        assert_equal("unknown", Magic.string_charset("over, and over, and over\\#{windows_char.to_s(8)}"))
      end
    EOMETHOD
  end

  # although this is redundant, it's nice to document the one weird case we've found explicitly
  def test_string_charset_for_string_with_windows_ellipsis_returns_unknown
    # windows 1252 ellipsis is 133 = 0205
    # this is a weird case found from a file Sylvia had problem slurping
    assert_equal("unknown", Magic.string_charset("over, and over, and over\205"))
  end

  def test_string_charset_for_string_with_utf8_Angstrom_returns_utf8_not_unknown
    string = File.open(absolute_path("utf-8.csv.gz")) do |io|
      uncompressed_io = Zlib::GzipReader.new(io)
      uncompressed_io.read
    end
    assert_equal("utf-8", Magic.string_charset(string))
  end

  def test_io_charset_for_ascii_file
    assert_equal("us-ascii", Magic.io_charset(File.open((absolute_path("us-ascii.txt")))))
  end

  def test_io_charset_for_utf8_file
    assert_equal("utf-8", Magic.io_charset(File.open(absolute_path("utf-8.txt"))))
  end

  require "zlib"
  def test_io_charset_for_gzipped_utf8_file
    File.open(absolute_path("utf-8.csv.gz")) do |io|
      uncompressed_io = Zlib::GzipReader.new(io)
      assert_equal("utf-8", Magic.io_charset(uncompressed_io))
    end
  end

  def test_io_charset_for_iso_8859_1_file
    assert_equal("iso-8859-1", Magic.io_charset(File.open(absolute_path("iso-8859-1.txt"))))
  end

  def test_io_charset_for_windows_1252_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.io_charset(File.open(absolute_path("windows-1252.txt"))))
  end

  def test_io_charset_for_macintosh_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.io_charset(File.open(absolute_path("macintosh.txt"))))
  end

  def test_io_charset_for_csv_file_that_looked_like_ppm_image
    assert_equal("us-ascii", Magic.io_charset(File.open(absolute_path("file_with_text_that_looked_like_ppm_image.csv"))))
  end

  def test_file_charset_for_ascii_file
    assert_equal("us-ascii", Magic.file_charset(absolute_path("us-ascii.txt")))
  end

  def test_file_charset_for_large_CSV_file_that_libmagic_thinks_is_pascal_sourcecode
    assert_equal("us-ascii", Magic.file_charset(absolute_path("part_of_ki_file.csv")))
  end

  def test_file_charset_for_utf8_file
    assert_equal("utf-8", Magic.file_charset(absolute_path("utf-8.txt")))
  end

  def test_file_charset_for_iso_8859_1_file
    assert_equal("iso-8859-1", Magic.file_charset(absolute_path("iso-8859-1.txt")))
  end

  def test_file_charset_for_windows_1252_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.file_charset(absolute_path("windows-1252.txt")))
  end

  def test_file_charset_for_macintosh_file
    # unfortunately, unknown means some kind of extended ascii
    assert_equal("unknown", Magic.file_charset(absolute_path("macintosh.txt")))
  end

  def test_file_charset_for_csv_file_that_looked_like_ppm_image
    assert_equal("us-ascii", Magic.file_charset(absolute_path("file_with_text_that_looked_like_ppm_image.csv")))
  end

  def test_file_charset_raises_if_file_does_not_exist
    # for this, we don't use assert_raise
    begin
      Magic.file_charset("some file that does not exist.txt")
      fail "Did not raise"
    rescue Exception => expected
      # ruby 1.9 and 1.8 return different exceptions
      assert(expected.message =~ /(some file that does not exist.txt|NULL pointer)/i)
    end
  end

  def test_file_charset_bang_exhaustively_checks_file_contents
    t1 = Time.now
    assert_equal("iso-8859-1", Magic.file_charset!(absolute_path("huge_file_with_one_special_character.csv")))
    puts "took #{Time.now - t1} seconds"
  end

  def test_file_charset_bang_returns_correct_value_for_us_ascii_file
    assert_equal("us-ascii", Magic.file_charset!(absolute_path("us-ascii.txt")))
  end

  def test_file_charset_bang_returns_correct_value_for_windows_1252_file
    assert_equal("unknown", Magic.file_charset!(absolute_path("windows-1252.txt")))
  end

  def test_file_charset_bang_returns_correct_value_for_UTF8_file
    assert_equal("utf-8", Magic.file_charset!(absolute_path("utf-8.txt")))
  end

  def test_file_charset_bang_handles_special_character_at_the_end_of_the_file
    assert_equal("us-ascii", Magic.file_charset(absolute_path("huge_file_with_one_special_character_at_the_end.csv")))
    assert_equal("iso-8859-1", Magic.file_charset!(absolute_path("huge_file_with_one_special_character_at_the_end.csv")))
  end

  def test_collect_special_characters_is_empty_when_there_are_no_special_characters
    assert_special_chars_equal("", "")
    assert_special_chars_equal("", "hello")
    assert_special_chars_equal("", "12345678901234567890")
  end

  def test_collect_special_characters_returns_characters_with_context
    assert_special_chars_equal("µ", "µ")
    assert_special_chars_equal("321µ123", "321µ123")
    assert_special_chars_equal("µ123", "µ123")
    assert_special_chars_equal("321µ", "321µ")
    assert_special_chars_equal("0987654321\xC21234567890", "0987654321\xC21234567890")
    assert_special_chars_equal("0987654321µ1234567890", "XXX0987654321µ1234567890XXX")
    assert_special_chars_equal("µ1234567890", "µ1234567890XXX")
    assert_special_chars_equal("0987654321µ", "XXX0987654321µ")
  end

  def test_collect_special_characters_does_not_duplicate_context
    assert_special_chars_equal("0987654321µaaaaaµ1234567890", "XXX0987654321µaaaaaµ1234567890XXX")
  end

  def test_collect_special_characters_works_with_multiple_characters
    assert_special_chars_equal(
      "0987654321µaaaaaµ12345678900987654321µ1234567890",
      "XXX0987654321µaaaaaµ1234567890XXXXXX0987654321µ1234567890XXX"
    )
  end

  def test_collect_special_characters_works_when_reading_multiple_chunks_to_the_buffer
    default_chunk_size = 2 ** 15
    assert_equal(default_chunk_size, Magic::CHUNK_SIZE)
    begin
      Magic.send(:remove_const, "CHUNK_SIZE")
      Magic.send(:const_set, "CHUNK_SIZE", 2)
      assert_special_chars_equal("µ", "µ")
      assert_special_chars_equal("µ123", "µ123")
      assert_special_chars_equal("321µ", "321µ")
      assert_special_chars_equal("321µ123", "321µ123")
      assert_special_chars_equal("0987654321µ", "XXX0987654321µ")
      assert_special_chars_equal(
        "0987654321µaaaaaµ12345678900987654321µ1234567890",
        "XXX0987654321µaaaaaµ1234567890XXXXXX0987654321µ1234567890XXX"
      )
    ensure
      Magic.send(:remove_const, "CHUNK_SIZE") if Magic.const_defined?("CHUNK_SIZE")
      Magic.const_set("CHUNK_SIZE", default_chunk_size)
    end
    assert_equal(default_chunk_size, Magic::CHUNK_SIZE)
  end

  def assert_special_chars_equal(expected_output, input)
    assert_equal(expected_output.force_encoding(Encoding::BINARY), Magic.send(:collect_special_characters, StringIO.new(input)))
  end

  def absolute_path(test_file_name)
    "#{ENV["PWD"]}/test/files/#{test_file_name}"
  end
end


ext/libmagic_wrapper.c
#include "ruby.h"
#include <magic.h>

VALUE rb_cMagic;
VALUE magic_file_mime_type(VALUE instance, VALUE str);
VALUE magic_string_mime_type(VALUE instance, VALUE str);
VALUE process_result(magic_t cookie, const char *mime_type);
magic_t load_cookie();

magic_t load_cookie()
{
    magic_t cookie = magic_open(MAGIC_MIME | MAGIC_ERROR);
    magic_load(cookie, NULL);
    return cookie;
}

VALUE process_result(magic_t cookie, const char *mime_type)
{
    if (mime_type == NULL) {
        rb_raise(rb_eRuntimeError, magic_error(cookie));
    } else {
        return rb_str_new(mime_type, strlen(mime_type));
    }
}

VALUE magic_file_mime_type(VALUE instance, VALUE str) 
{
    const char *filename = RSTRING(str)->ptr;

    magic_t cookie = load_cookie();
    const char *mime_type = magic_file(cookie, filename);
    return process_result(cookie, mime_type);
}

VALUE magic_string_mime_type(VALUE instance, VALUE str)
{
    const char *c_str = RSTRING(str)->ptr;
    
    magic_t cookie = load_cookie();
    const char *mime_type = magic_buffer(cookie, c_str, strlen(c_str));
    return process_result(cookie, mime_type);
}

void Init_libmagic_wrapper (void) {
    rb_cMagic = rb_define_class("Magic", rb_cObject);
    rb_define_singleton_method(rb_cMagic, "file_mime_type", magic_file_mime_type, 1);
    rb_define_singleton_method(rb_cMagic, "string_mime_type", magic_string_mime_type, 1);
}


pkg/libmagic-0.5.11/lib/libmagic.rb
require 'ffi'

module Magic
  VERSION = '0.5.11'
  ASCII_CHARSET = "us-ascii"
  # currently libmagic doesn't distinguish the various extended ASCII charsets except ISO-8859-1
  EXTENDED_ASCII_CHARSET = "unknown"
  PROBLEMATIC_EXTENDED_ASCII_CHAR = 133 # windows-1252 ellipsis
  REGEX = /charset=(.+)$/
  CHUNK_SIZE = 2 ** 15

  class << self
    def file_charset(filename)
      mime_type_to_charset(file_mime_type(filename))
    end

    # Exhaustively checks file contents. The plan is you should always be able to trust the answer this method returns.
    def file_charset!(filename)
      quick_answer = file_charset(filename)

      if quick_answer == ASCII_CHARSET
        return File.open(filename) { |io| io_charset(io) } # try harder
      else
        return quick_answer
      end
    end

    def io_charset(io)
      io.rewind
      special_characters = collect_special_characters(io)
      return special_characters.empty? ? ASCII_CHARSET : string_charset(special_characters)
    ensure
      io.rewind
    end

    def string_charset(text)
      quick_answer = mime_type_to_charset(string_mime_type(text))
      if quick_answer == ASCII_CHARSET
        text.each_byte { |byte| return EXTENDED_ASCII_CHARSET if byte == PROBLEMATIC_EXTENDED_ASCII_CHAR }
      end
      return quick_answer
    end

    private
    CONTEXT_SIZE = 10
    LAST_ASCII_CHAR = 127
    def collect_special_characters(io)
      special_characters_with_context = ""
      buffer = ""
      leading_index = trailing_index = buffer_bytesize = 0 # cache buffer.bytesize for performance
      last_detection = nil

      while leading_index < buffer_bytesize || !io.eof? do
        # Add to buffer if needed
        if leading_index == buffer_bytesize
          # Need to read more, but save whatever is after trailing_index in the current buffer
          buffer.slice!(0...trailing_index)
          # adjust all indices
          last_detection -= trailing_index if last_detection
          leading_index  -= trailing_index
          trailing_index -= trailing_index
          buffer << io.read(CHUNK_SIZE)
          buffer_bytesize = buffer.bytesize
        end

        byte = buffer.getbyte(leading_index)
        if byte > LAST_ASCII_CHAR
          last_detection = leading_index
        end

        if last_detection
          if last_detection < (leading_index - CONTEXT_SIZE)
            # It has been CONTEXT_SIZE bytes since the last non-ascii character, so we should write this chunk to the results
            special_characters_with_context << buffer.byteslice(trailing_index...leading_index)
            trailing_index = leading_index + 1 # Just think of it as trailing_index = "leading_index's value at the end of the loop"
            last_detection = nil
          end
        else
          # Advance the trailing index if it we haven't seen a non-ascii byte and the trailing index is CONTEXT_SIZE bytes behind
          if trailing_index == leading_index - CONTEXT_SIZE
            trailing_index += 1
          end
        end

        leading_index += 1
      end

      # Deal with leftovers, if present
      if last_detection
        special_characters_with_context << buffer.byteslice(trailing_index, leading_index)
      end

      return special_characters_with_context
    end

    def mime_type_to_charset(mime_type)
      if match = REGEX.match(mime_type)
        return standardize_charset(match[1])
      else
        return nil
      end
    end

    # file returns different things in different versions
    def standardize_charset(nonstandard)
      case nonstandard
      when "unknown-8bit"
        "unknown"
      else
        nonstandard
      end
    end
  end
end

# Just to make things neater, split out the FFI part here
module Magic
  extend FFI::Library

  ffi_lib "magic" # you might need to set your LD_LIBRARY_PATH on OS X if you're using MacPorts

  private
  MAGIC_NONE =              0x000000 # No flags
  MAGIC_DEBUG =             0x000001 # Turn on debugging
  MAGIC_SYMLINK =           0x000002 # Follow symlinks
  MAGIC_COMPRESS =          0x000004 # Check inside compressed files
  MAGIC_DEVICES =           0x000008 # Look at the contents of devices
  MAGIC_MIME_TYPE =         0x000010 # Return only the MIME type
  MAGIC_CONTINUE =          0x000020 # Return all matches
  MAGIC_CHECK =             0x000040 # Print warnings to stderr
  MAGIC_PRESERVE_ATIME =    0x000080 # Restore access time on exit
  MAGIC_RAW =               0x000100 # Don't translate unprint chars
  MAGIC_ERROR =             0x000200 # Handle ENOENT etc as real errors
  MAGIC_MIME_ENCODING =     0x000400 # Return only the MIME encoding
  MAGIC_MIME =              (MAGIC_MIME_TYPE | MAGIC_MIME_ENCODING)
  MAGIC_NO_CHECK_COMPRESS = 0x001000 # Don't check for compressed files
  MAGIC_NO_CHECK_TAR =      0x002000 # Don't check for tar files
  MAGIC_NO_CHECK_SOFT =     0x004000 # Don't check magic entries
  MAGIC_NO_CHECK_APPTYPE =  0x008000 # Don't check application type
  MAGIC_NO_CHECK_ELF =      0x010000 # Don't check for elf details
  MAGIC_NO_CHECK_ASCII =    0x020000 # Don't check for ascii files
  MAGIC_NO_CHECK_TROFF =    0x040000 # Don't check ascii/troff
  MAGIC_NO_CHECK_TOKENS =   0x100000 # Don't check ascii/tokens

  attach_function :magic_open, [:int], :pointer
  attach_function :magic_setflags, [:pointer, :int], :int
  attach_function :magic_load, [:pointer, :pointer], :int
  attach_function :magic_buffer, [:pointer, :pointer, :size_t], :string
  attach_function :magic_file, [:pointer, :string], :string
  attach_function :magic_error, [:pointer], :string

  class << self
    def string_mime_type(string)
      cookie = load_cookie
      return process_result(cookie, magic_buffer(cookie, string, string.size))
    end

    def file_mime_type(filename)
      cookie = load_cookie
      return process_result(cookie, magic_file(cookie, filename))
    end

    private
    def load_cookie
      # MAGIC_NO_CHECK_SOFT is supposed to prevent it from using the magic database, but doesn't seem to work on Linux.
      cookie = magic_open(MAGIC_MIME | MAGIC_ERROR)
      # Instead, we use a custom magic file with one simple entry.
      magic_load(cookie, "#{File.dirname(__FILE__)}/custom-magic")
      return cookie
    end

    def process_result(cookie, mime_type)
      return mime_type unless mime_type.nil?
      raise magic_error(cookie)
    end
  end
end
